controllers/trainingjob/trainingjob_controller.go

```diff
@@ -226,6 +226,7 @@ func (r *Reconciler) initializeContext(ctx *reconcileRequestContext) error {
 		ctx.TrainingJobName = *ctx.TrainingJob.Spec.TrainingJobName
 	} else {
 		ctx.TrainingJobName = controllers.GetGeneratedJobName(ctx.TrainingJob.ObjectMeta.GetUID(), ctx.TrainingJob.ObjectMeta.GetName(), MaxTrainingJobNameLength)
+		ctx.TrainingJob.Spec.TrainingJobName = &ctx.TrainingJobName
```

Comment:
We should call `r.client.Update` here, to guarantee that the spec name is persisted to etcd.